lib/elixir/lib/list.ex

```diff
@@ -577,22 +577,51 @@ defmodule List do
   end
 
   @doc """
-  Returns `true` if `list` starts with the given `prefix` list.
+  Returns `true` if `list` starts with the given `prefix` list; otherwise returns `false`.
+
+  `list` and `prefix` can be proper or improper lists.
+
+  If `prefix` is an empty list, it returns `true`.
 
   ### Examples
 
       iex> List.starts_with?([1, 2, 3], [1, 2])
       true
+
+      iex> List.starts_with?([1, 2], [1, 2, 3])
+      false
+
       iex> List.starts_with?([:alpha], [])
       true
+
       iex> List.starts_with?([], [:alpha])
       false
 
+      iex> List.starts_with?([1, 2 | 3], [1, 2])
+      true
+
+      iex> List.starts_with?([1, 2 | 3], [1, 2 | 3])
+      true
+
+      iex> List.starts_with?([1, 2, 3], [1, 2 | 3])
+      false
+
   """
-  @spec starts_with?(list, list) :: boolean
-  def starts_with?(list, prefix) do
-    :lists.prefix(prefix, list)
-  end
+  @spec starts_with?(maybe_improper_list, maybe_improper_list) :: boolean
+  @spec starts_with?(maybe_improper_list, []) :: true
+  @spec starts_with?([], nonempty_maybe_improper_list) :: false
+  def starts_with?(list, prefix)
+  def starts_with?(list, list) when is_list(list),
```

Comment:
Since `:lists.prefix/2` already takes care of improper lists by raising a (not very friendly but still) `FunctionClauseError`, I don't really see the point of implementing our own unless it's to substantially improve the error message.